File: cg/resolverpt2.py
import rclpy
from rclpy.node import Node
from cg_interfaces.srv import MoveCmd, Reset
from cg_interfaces.msg import RobotSensors
from collections import deque
import time

TARGET_POS = (14, 14)

class MazeResolverpt2(Node):
    def __init__(self):
        super().__init__('maze_resolverpt2')

        self.move_client = self.create_client(MoveCmd, '/move_command')
        self.reset_client = self.create_client(Reset, '/reset')
        
        self.sensor_sub = self.create_subscription(
            RobotSensors, 
            '/culling_games/robot_sensors', 
            self.sensor_callback, 
            10
        )
        
        self.latest_sensors = None
        self.robot_pos = [1, 1]
        self.stack = []
        self.visited = set()
        
        self.rows, self.cols = 29, 29
        self.internal_map = [['?' for _ in range(self.cols)] for _ in range(self.rows)]
        self.internal_map[1][1] = 'f'

        while not self.move_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Aguardando move_command...')
        while not self.reset_client.wait_for_service(timeout_sec=1.0):
            self.get_logger().info('Aguardando reset...')

    def sensor_callback(self, msg):
        self.latest_sensors = msg

    def reset_game(self):
        req = Reset.Request()
        req.is_random = True
        future = self.reset_client.call_async(req)
        rclpy.spin_until_future_complete(self, future)
        return future.result()

    def move_robot(self, direction):
        req = MoveCmd.Request()
        req.direction = direction
        future = self.move_client.call_async(req)
        rclpy.spin_until_future_complete(self, future)
        res = future.result()
        if res.success:
            self.robot_pos = list(res.robot_pos)
        return res.success

    def update_map_from_sensors(self):
        if not self.latest_sensors:
            return

        r, c = self.robot_pos
        sensors = self.latest_sensors
        
        deltas = {
            'up': (-1, 0, sensors.up),
            'down': (1, 0, sensors.down),
            'left': (0, -1, sensors.left),
            'right': (0, 1, sensors.right)
        }

        for direction, (dr, dc, val) in deltas.items():
            nr, nc = r + dr, c + dc
            if 0 <= nr < self.rows and 0 <= nc < self.cols:
                if self.internal_map[nr][nc] == '?':
                    self.internal_map[nr][nc] = val

    def bfs_check(self, start, target):
        """Valida o mapa gerado encontrando um caminho nela"""
        queue = deque([(tuple(start), [])])
        visited_bfs = set([tuple(start)])
        
        # Heurística de movimento para o BFS também
        directions = [(1, 0, 'down'), (0, 1, 'right'), (-1, 0, 'up'), (0, -1, 'left')]

        while queue:
            (curr_r, curr_c), path = queue.popleft()
            if (curr_r, curr_c) == tuple(target):
                return path

            for dr, dc, move_name in directions:
                nr, nc = curr_r + dr, curr_c + dc
                if 0 <= nr < self.rows and 0 <= nc < self.cols:
                    cell = self.internal_map[nr][nc]
                    # Só anda se for chão conhecido
                    if cell in ['f', 't', 'r'] and (nr, nc) not in visited_bfs:
                        visited_bfs.add((nr, nc))
                        queue.append(((nr, nc), path + [move_name]))
        return None

    def run(self):
        self.get_logger().info('Iniciando Exploração/mapeamento de área ...')
        res = self.reset_game()
        self.get_logger().info(f'Mapa Aleatório Loaded: {res.loaded_map_name}')
        
        # Espera inicial pelos sensores
        while self.latest_sensors is None:
            rclpy.spin_once(self, timeout_sec=0.1)
        
        self.visited.add(tuple(self.robot_pos))
        self.stack.append(tuple(self.robot_pos))
        
        target_found_pos = None

        while self.stack:
            self.latest_sensors = None
            while self.latest_sensors is None:
                rclpy.spin_once(self, timeout_sec=0.1)
            
            self.update_map_from_sensors()
            curr_r, curr_c = self.robot_pos

            if self.internal_map[curr_r][curr_c] == 't':
                target_found_pos = (curr_r, curr_c)
                break
                
            nearby_target = False
            for dr, dc, move in [(1,0,'down'),(0,1,'right'),(-1,0,'up'),(0,-1,'left')]:
                 nr, nc = curr_r + dr, curr_c + dc
                 if 0 <= nr < self.rows and 0 <= nc < self.cols:
                     if self.internal_map[nr][nc] == 't':
                         self.move_robot(move)
                         target_found_pos = (nr, nc)
                         nearby_target = True
                         break
            if nearby_target:
                break

            valid_moves = []
            possible_directions = [('down', 1, 0), ('right', 0, 1), ('up', -1, 0), ('left', 0, -1)]
            
            for move_name, dr, dc in possible_directions:
                nr, nc = curr_r + dr, curr_c + dc
                
                if 0 <= nr < self.rows and 0 <= nc < self.cols:
                    cell_val = self.internal_map[nr][nc]
                    
                    if cell_val in ['f', 't'] and (nr, nc) not in self.visited:

                        dist_to_center = abs(nr - TARGET_POS[0]) + abs(nc - TARGET_POS[1])
                        

                        valid_moves.append((dist_to_center, move_name, dr, dc, nr, nc))
            
            valid_moves.sort(key=lambda x: x[0])

            if valid_moves:
                _, best_move, _, _, nr, nc = valid_moves[0]
                
                if self.move_robot(best_move):
                    self.visited.add((nr, nc))
                    self.stack.append((curr_r, curr_c))
                    time.sleep(0.05)
                else:
                    pass
            else:
                if len(self.stack) > 0:
                    back_r, back_c = self.stack.pop()
                    dr = back_r - curr_r
                    dc = back_c - curr_c
                    
                    back_move = ''
                    if dr == 1: back_move = 'down'
                    elif dr == -1: back_move = 'up'
                    elif dc == 1: back_move = 'right'
                    elif dc == -1: back_move = 'left'
                    
                    if back_move:
                        self.move_robot(back_move)
                        time.sleep(0.05)
                else:
                    break

        if target_found_pos:
            self.get_logger().info("--- ALVO ENCONTRADO! ---")
            self.get_logger().info("Comprovando mapa com BFS...")
            path = self.bfs_check((1,1), target_found_pos)
            if path:
                self.get_logger().info(f"SUCESSO! Mapa válido. Rota de {len(path)} passos calculada.")
            else:
                self.get_logger().error("ERRO: Falha ao calcular rota no mapa gerado.")
        else:
            self.get_logger().warn("Exploração finalizada sem encontrar o alvo.")

def main(args=None):
    rclpy.init(args=args)
    resolverpt2 = MazeResolverpt2()
    resolverpt2.run()
    resolverpt2.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: cg_solution/src/resolverpt2.cpp
#include <rclcpp/rclcpp.hpp>
#include <cg_interfaces/srv/move_cmd.hpp>
#include <cg_interfaces/srv/reset.hpp>
#include <cg_interfaces/msg/robot_sensors.hpp>
#include <deque>
#include <vector>
#include <string>
#include <set>
#include <thread>
#include <algorithm>
#include <iostream>
#include <cmath>
#include <tuple>

using namespace std;
using namespace rclcpp;


const int ALVO_R = 14;
const int ALVO_C = 14;

class MazeResolverPt2 : public Node {
public:
    Client<cg_interfaces::srv::MoveCmd>::SharedPtr move_cli;
    Client<cg_interfaces::srv::Reset>::SharedPtr reset_cli;
    Subscription<cg_interfaces::msg::RobotSensors>::SharedPtr sensor_sub;
    
    cg_interfaces::msg::RobotSensors::SharedPtr last_sensor;
    
    string my_map[29][29];
    int r_pos, c_pos;

    MazeResolverPt2() : Node("maze_resolverpt2") {
        move_cli = create_client<cg_interfaces::srv::MoveCmd>("/move_command");
        reset_cli = create_client<cg_interfaces::srv::Reset>("/reset");
        
        sensor_sub = create_subscription<cg_interfaces::msg::RobotSensors>(
            "/culling_games/robot_sensors", 10,
            [this](cg_interfaces::msg::RobotSensors::SharedPtr msg){
                this->last_sensor = msg;
            });
            
        for(int i=0; i<29; i++) for(int j=0; j<29; j++) my_map[i][j] = "?";
        my_map[1][1] = "f";
        r_pos = 1; c_pos = 1;
        
        while(!move_cli->wait_for_service(1s)) cout << ".";
        while(!reset_cli->wait_for_service(1s)) cout << ".";
        cout << " Ready!" << endl;
    }

    bool call_move(string d) {
        auto req = make_shared<cg_interfaces::srv::MoveCmd::Request>();
        req->direction = d;
        auto f = move_cli->async_send_request(req);
        spin_until_future_complete(this->shared_from_this(), f);
        auto res = f.get();
        if(res->success) {
            r_pos = res->robot_pos[0];
            c_pos = res->robot_pos[1];
        }
        return res->success;
    }

    void update_sensors() {
        if(!last_sensor) return;
        if(r_pos > 0 && my_map[r_pos-1][c_pos] == "?") my_map[r_pos-1][c_pos] = last_sensor->up;
        if(r_pos < 28 && my_map[r_pos+1][c_pos] == "?") my_map[r_pos+1][c_pos] = last_sensor->down;
        if(c_pos > 0 && my_map[r_pos][c_pos-1] == "?") my_map[r_pos][c_pos-1] = last_sensor->left;
        if(c_pos < 28 && my_map[r_pos][c_pos+1] == "?") my_map[r_pos][c_pos+1] = last_sensor->right;
    }

    vector<string> validate_map(int tr, int tc) {
        queue<pair<pair<int,int>, vector<string>>> q;
        set<pair<int,int>> vis;
        q.push({{1,1}, {}});
        vis.insert({1,1});
        
        int dr[]={1, 0, -1, 0}; 
        int dc[]={0, -1, 0, 1};
        string n[]={"down", "left", "up", "right"};
        
        while(!q.empty()){
            auto curr = q.front(); q.pop();
            int r=curr.first.first; int c=curr.first.second;
            if(r==tr && c==tc) return curr.second;
            
            for(int i=0;i<4;i++){
                int nr=r+dr[i]; int nc=c+dc[i];
                if(nr>=0 && nr<29 && nc>=0 && nc<29){
                    string val = my_map[nr][nc];
                    if((val=="f"||val=="t"||val=="r") && vis.count({nr,nc})==0){
                        vis.insert({nr,nc});
                        vector<string> p=curr.second; p.push_back(n[i]);
                        q.push({{nr,nc}, p});
                    }
                }
            }
        }
        return {};
    }

    void run_logic() {
        auto r_req = make_shared<cg_interfaces::srv::Reset::Request>();
        r_req->is_random = true;
        auto rf = reset_cli->async_send_request(r_req);
        spin_until_future_complete(this->shared_from_this(), rf);
        cout << "Mapa: " << rf.get()->loaded_map_name << endl;

        while(!last_sensor) { spin_some(this->shared_from_this()); this_thread::sleep_for(100ms); }

        vector<pair<int,int>> stack;
        set<pair<int,int>> visited;
        
        visited.insert({r_pos, c_pos});
        stack.push_back({r_pos, c_pos});

        int final_r = -1, final_c = -1;
        bool found = false;

        while(!stack.empty()) {
            last_sensor = nullptr;
            while(!last_sensor) { spin_some(this->shared_from_this()); this_thread::sleep_for(50ms); }
            update_sensors();

            if(my_map[r_pos][c_pos] == "t") {
                final_r=r_pos; final_c=c_pos; found=true; break;
            }


            int drs[]={1, 0, -1, 0}; 
            int dcs[]={0, -1, 0, 1}; 
            string nms[]={"down", "left", "up", "right"};
            
            for(int i=0;i<4;i++){
                int nr=r_pos+drs[i]; int nc=c_pos+dcs[i];
                if(nr>=0 && nr<29 && nc>=0 && nc<29){
                    if(my_map[nr][nc]=="t"){
                        call_move(nms[i]);
                        final_r=nr; final_c=nc; found=true; vizinho_alvo=true;
                        break;
                    }
                }
            }
            if(vizinho_alvo) break;

            vector<tuple<int, string, int, int, int, int>> valid_moves;
            
            for(int i=0; i<4; i++) {
                int nr = r_pos + drs[i];
                int nc = c_pos + dcs[i];
                if(nr>=0 && nr<29 && nc>=0 && nc<29) {
                    string cell = my_map[nr][nc];
                    if((cell=="f" || cell=="t") && visited.count({nr, nc}) == 0) {
                        
                        int dist_base = abs(nr - ALVO_R) + abs(nc - ALVO_C);
                        
                        int penalidade = 0;
                        if(nms[i] == "down") penalidade = 0;      
                        else if(nms[i] == "left") penalidade = 1; 
                        else if(nms[i] == "up") penalidade = 2;
                        else if(nms[i] == "right") penalidade = 3;
                        
                        int score_final = (dist_base * 10) + penalidade;

                        valid_moves.push_back({score_final, nms[i], drs[i], dcs[i], nr, nc});
                    }
                }
            }

            sort(valid_moves.begin(), valid_moves.end());

            if(valid_moves.size() > 0) {
                auto best = valid_moves[0];
                string best_move = get<1>(best); 
                int nr = get<4>(best); 
                int nc = get<5>(best); 
                
                int dr_used = get<2>(best);
                int dc_used = get<3>(best);

                if(call_move(best_move)) {
                    visited.insert({nr, nc});
                    stack.push_back({r_pos-dr_used, c_pos-dc_used}); 
                    this_thread::sleep_for(50ms);
                }
            } else {
                if(!stack.empty()) {
                    auto back_pos = stack.back();
                    stack.pop_back();
                    
                    int diff_r = back_pos.first - r_pos;
                    int diff_c = back_pos.second - c_pos;
                    
                    string b_move = "";
                    if(diff_r == -1) b_move = "down";
                    else if(diff_r == 1) b_move = "up";
                    else if(diff_c == 1) b_move = "right";
                    else if(diff_c == -1) b_move = "left";
                    
                    if(b_move != "") {
                        call_move(b_move);
                        this_thread::sleep_for(50ms);
                    }
                } else break;
            }
        }

        if(found) {
            cout << "--- ALVO ENCONTRADO! ---" << endl;
            auto path = validate_map(final_r, final_c);
            if(path.size() > 0) cout << "SUCESSO! Mapa valido. Passos: " << path.size() << endl;
            else cout << "ERRO: Caminho nao bate." << endl;
        } else {
            cout << "Desisto, nao achei." << endl;
        }
    }
};

int main(int argc, char **argv) {
    init(argc, argv);
    auto n = make_shared<MazeResolverPt2>();
    n->run_logic();
    shutdown();
    return 0;
}